File: src/adapters/grype.ts
import type { NormalizedVulnerability } from '../types/index.js';
import { mapSeverity } from './severity.js';

export function extractGrypeVulnerabilities(report: any): NormalizedVulnerability[] {
  const findings: NormalizedVulnerability[] = [];

  if (report?.matches) {
    for (const match of report.matches) {
      const vuln = match.vulnerability;
      findings.push({
        source: 'grype',
        cveId: vuln.id,
        packageName: match.artifact.name,
        installedVersion: match.artifact.version ?? undefined,
        fixedVersion: vuln.fix?.versions?.[0] ?? undefined,
        severity: mapSeverity(vuln.severity) as NormalizedVulnerability['severity'],
        cvssScore: vuln.cvss?.[0]?.metrics?.baseScore ?? undefined,
        vulnerabilityUrl: vuln.urls?.[0] ?? undefined,
        description: vuln.description ?? undefined,
      });
    }
  }

  return findings;
}

File: internal/adapter/grype.go
package adapter

import (
	"github.com/HarborGuard/harborguard-sensor/internal/types"
)

// ExtractGrypeVulnerabilities extracts normalized vulnerabilities from grype JSON.
func ExtractGrypeVulnerabilities(report interface{}) []types.NormalizedVulnerability {
	var findings []types.NormalizedVulnerability

	data, ok := report.(map[string]interface{})
	if !ok {
		return findings
	}

	matches, ok := data["matches"].([]interface{})
	if !ok {
		return findings
	}

	for _, m := range matches {
		match, ok := m.(map[string]interface{})
		if !ok {
			continue
		}

		vuln, _ := match["vulnerability"].(map[string]interface{})
		artifact, _ := match["artifact"].(map[string]interface{})
		if vuln == nil || artifact == nil {
			continue
		}

		var fixedVersion string
		if fix, ok := vuln["fix"].(map[string]interface{}); ok {
			if versions, ok := fix["versions"].([]interface{}); ok && len(versions) > 0 {
				if v, ok := versions[0].(string); ok {
					fixedVersion = v
				}
			}
		}

		var cvssScore *float64
		if cvssArr, ok := vuln["cvss"].([]interface{}); ok && len(cvssArr) > 0 {
			if first, ok := cvssArr[0].(map[string]interface{}); ok {
				if metrics, ok := first["metrics"].(map[string]interface{}); ok {
					if s, ok := metrics["baseScore"].(float64); ok {
						cvssScore = &s
					}
				}
			}
		}

		var vulnURL string
		if urls, ok := vuln["urls"].([]interface{}); ok && len(urls) > 0 {
			if u, ok := urls[0].(string); ok {
				vulnURL = u
			}
		}

		findings = append(findings, types.NormalizedVulnerability{
			Source:           "grype",
			CveID:            getString(vuln, "id"),
			PackageName:      getString(artifact, "name"),
			InstalledVersion: getString(artifact, "version"),
			FixedVersion:     fixedVersion,
			Severity:         MapSeverity(getString(vuln, "severity")),
			CvssScore:        cvssScore,
			VulnerabilityURL: vulnURL,
			Description:      getString(vuln, "description"),
		})
	}

	return findings
}
